crypto/modes/siv128.c

```diff
@@ -0,0 +1,365 @@
+/*
+ * Copyright 2017 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the OpenSSL license (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <string.h>
+#include <stdlib.h>
+#include <openssl/crypto.h>
+#include <openssl/cmac.h>
+#include "modes_lcl.h"
+
+#ifndef OPENSSL_NO_SIV
+
+# if !defined(BSWAP8)
+#  if defined(__GNUC__) || defined(__clang__)
+__owur static ossl_inline uint64_t BSWAP8(uint64_t x)
+{
+    return __builtin_bswap64(x);
+}
+#  elif defined(_MSC_VER)
+__owur static ossl_inline uint64_t BSWAP8(uint64_t x)
+{
+    return _byteswap_uint64(x);
```

Comment:
Well, same question here, but then also question is if it takes even additional #pragma intrinsic. I mean there surely is _byteswap_uint64 in library, but that's not really what we want, is it?

Reply:
It seems that /O2 implies pragma intrinsic...

Reply:
Another thing to keep in mind is that _MSC_VER is not exclusive to x86.

> there surely is _byteswap_uint64 in library, but that's not really what we want, is it?

I mean that if given a choice between calling library and doing byte swap ourselves, we probably should choose the latter.

Reply:
_byteswap_uint64() is available all the way back to VS 2005, all versions before that are "retired".

Reply:
So... I did some timings on the intrinsic, this version, and another byteswap implementation, and this one is consistently faster... so, I'll remove the others and just do our own (although, it could be done in two x86-64 instructions...)